Add Keeper Cron contract (#598)
smartcontractkit/documentation: docs/chainlink-keepers/cron-contracts.md

---
layout: nodes.liquid
section: ethereum
date: Last Modified
title: 'Keepers Job Scheduler [BETA]'
whatsnext:
  {
    'FAQs': '/docs/chainlink-keepers/faqs/',
  }
---

> 👍 Beta
>
> The Keepers Job Scheduler is currently in Beta.

# Overview

> In this tutorial you will use the Chainlink Keepers Job Scheduler to set up a time schedule to automatically execute smart contract functions.

Reviewer: ```suggestion
In this tutorial you will use the Chainlink Keepers Job Scheduler to automatically execute smart contract functions on a specific time schedule.
```
This might read more smoothly? Feel free to reject this suggestion.

**Table of Contents**

+ [Overview](#overview)
+ [What is Job Scheduler?](#what-is-job-scheduler)
+ [Job Scheduling Tutorial](#job-scheduling-tutorial)
+ [Resources](#resources)

# What is Job Scheduler?

The Chainlink Keepers Job Scheduler schedules jobs that use the Chainlink Keepers Network to automatically execute your smart contract functions at specific times. After you register your Job Scheduler with the [Chainlink Keepers network](https://keepers.chain.link/new), the decentralized Keepers network monitors and executes your scheduled jobs.  

# Job Scheduling Tutorial

> Use [Remix](https://remix.ethereum.org/) for this tutorial. To learn more about Remix and how you can use it to deploy a smart contract, visit our [tutorial](/docs/deploy-your-first-contract/).

Reviewer: ```suggestion
Use [Remix](https://remix.ethereum.org/) for this tutorial. To learn more about Remix and how you can use it to deploy a smart contract, see the [Getting Started guide](/docs/deploy-your-first-contract/).
```

```

```json
<your target address>, <your function specification>, <your encodedCronSpec>
```
Replace the values and paste the snippet above into the `createCronJobFromEncodedSpec` function in Remix and execute the function. Once executed, you can call the `getActiveCronJobIDs` to see if your job has been registered. You can execute the `getCronJob` function to view the details of your job played back. You can add multiple jobs on this contract in this fashion, and also use the other in your Job Scheduler to delete jobs or pause and unpause the Scheduler. However, you must register your Job Scheduler with the Chainlink Keepers network for it to be monitored and executed.

## Setting up Chainlink Keepers

> Because the deployed contract is already [Keeper-compatible](../compatible-contracts), you can immediately register it with Keepers following these [steps](../register-upkeep). Be sure to provide the address of your Job Scheduler contract in **Upkeep address**. For gas limits, you should specify the upper limit of gas your target function will use.

Reviewer: ```suggestion
Because the deployed contract is already [Keeper-compatible](../compatible-contracts), you can immediately [register the contract for upkeep(/docs/chainlink-keepers/register-upkeep/). Provide the address of your Job Scheduler contract in the **Upkeep address** field. For gas limits, specify the upper limit of gas that your target function will use.
```

> ❗️ Gas Limits
>
> The `KeeperRegistry` enforces a cap for gas used both on-chain and off-chain. See the [Keepers Network Overview](../overview/) for details. The caps are configurable and might change based on user feedback. Be sure that you understand these limits if your use case requires a large amount of gas.

> Once you have registered your contract, you can view your Upkeep. The *Upkeep History* will display **Perform Upkeep** for all registered jobs on your Job Scheduler contract. To ensure Chainlink Keepers monitors your Job Scheduler, please ensure you fund your Upkeep.

Reviewer: ```suggestion
After you register your contract, you can view your upkeep in the [Chainlink Keepers App](https://keepers.chain.link/). The *Upkeep History* displays **Perform Upkeep** events for all registered jobs on your Job Scheduler contract. To ensure that Chainlink Keepers continues to monitor your Job Scheduler, maintain a sufficient LINK balance on your Upkeep.
```

We recommend you first test the tutorial on an EVM-compatible testnet where Chainlink Keepers is deployed, before moving to mainnet.

## Creating your own Job Scheduler contract

To deploy your own on-chain Job Scheduler contract, we will use the externally audited and pre-deployed Chainlink  [CronUpkeepFactory](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/factories/CronUpkeepFactory.sol) smart contract. This is called the [Factory pattern](https://www.youtube.com/watch?v=Q1zZo4O_Ong), where one contract deploys another contract. To deploy your Job Scheduler (called `CronUpkeep`), select the new file icon in Remix, name it `CronUpkeepFactory.abi` and paste the [ABI](#cronupkeepfactory-abi) for `CronUpkeepFactory` from the [Resources](#resources) section below. The ABI is the application binary interface that is used to interact with the smart contract that has been deployed to the network.

![Remix ABI Cron](/images/keepers/cron-1.png)

> Now select the **DEPLOY & RUN TRANSACTIONS** icon in the left-hand navigation bar and set **Environment** to *Injected Web3*. Paste the address of our deployed `CronUpkeepFactory` contract in the **At Address** box and select the **At Address** button. You can find the [addresses](#contract-addresses) for deployed `CronUpkeepFactory` contracts in the [Resources](#resources) section. Once this step is completed, you will be able to view callable functions of the `CronUpkeepFactory` in Remix in the bottom left of the screen. Select the orange `newCronUpkeep` button and follow the prompts in MetaMask to confirm the transaction. When the transaction completes, you will see a transaction hash in the terminal pointing to the creation of your Job Scheduler.

Author: ```suggestion
Select the **DEPLOY & RUN TRANSACTIONS** icon in the left-hand navigation bar and set **Environment** to *Injected Web3*. Paste the address of our deployed `CronUpkeepFactory` contract in the **At Address** box and select the **At Address** button. You can find the [addresses](#contract-addresses) for deployed `CronUpkeepFactory` contracts in the [Resources](#resources) section. You will then be able to view callable functions of `CronUpkeepFactory`. Select the `newCronUpkeep` button and follow the prompts in MetaMask to confirm the transaction. When the transaction is complete,  a transaction hash will appear in the terminal. This points to the creation of your Job Scheduler.
```

## Finding the address of your Job Scheduler

> Please copy the transaction hash, and open it in the scanner of the chain you are working on. For example, if you were working on Ethereum Kovan please go to https://kovan.etherscan.io/ and search for your Transaction Hash. Now select the *Internal Txns* tab which will show the contract creation and contract address in the `To` field. Please note the address and transaction hash for future use.

Author: ```suggestion
Copy the transaction hash and open it in the scanner of the chain you are working on. Once you have found it, select the **Internal Txns** tab. This displays the contract creation and contract address in the `To` field. Note the address and transaction hash for future use.
```

![Remix Transaction Hash](/images/keepers/cron-2.png)

## Interacting with your Job Scheduler

> To interact with your Job Scheduler contract, create a new file called `CronUpkeep.abi` in Remix and paste the [ABI](#job-scheduler-aka-cronupkeep-abi) for `CronUpkeep` contract from the [Resources](#resources) section below. Select the **DEPLOY & RUN TRANSACTIONS** icon on the left and paste the address of your Job Scheduler contract in the **At Address** box, and then select the **At Address** button. Please accept the prompts after which the contract interface will open in the bottom left of Remix.

Author: ```suggestion
To interact with your Job Scheduler contract, create a new file called `CronUpkeep.abi` in Remix and paste the [ABI](#job-scheduler-aka-cronupkeep-abi) for `CronUpkeep` contract from the [Resources](#resources) section below. Select the **DEPLOY & RUN TRANSACTIONS** icon on the left and paste the address of your Job Scheduler contract in the **At Address** box, and then select the **At Address** button. Accept the prompts after which the contract interface will open in the bottom left of Remix.
```

![Remix Contract Interface](/images/keepers/cron-3.png)

## Scheduling a Job

To schedule a new job, use the function `createCronJobFromEncodedSpec`. This function has three inputs: 

+ The target address of the smart contract you want to automate
+ The [function signature](#function-signature), or the encoding of the function, you want to call 
+ The [encodedCronSpec](#encodedcronspec) encoding of your specified time interval.

> Please see the [Encodings](#encodings) section below to get the [function signature](#function-signature) and the [encodedCronSpec](#encodedcronspec) for your job. 

Author: ```suggestion
Refer to the [Encodings](#encodings) section below to retrieve the [function signature](#function-signature) and the [encodedCronSpec](#encodedcronspec) for your job. 
```

> We will now paste, minus <> brackets, 

Author: ```suggestion

```

```json
<your target address>, <your function specification>, <your encodedCronSpec>
```

> into the into the createCronJobFromEncodedSpec function in Remix and execute it. Once executed, you can call the getActiveCronJobIDs to see if your job has been registered. You can execute the getCronJob function to see the details of your job played back. You can add multiple jobs on this contract in this fashion, and also use the other functions in your Job Scheduler to delete jobs or pause and unpause the Scheduler.

Author: ```suggestion
Replace the values and paste the snippet above into the `createCronJobFromEncodedSpec` function in Remix and execute the function. Once executed, you can call the `getActiveCronJobIDs` to see if your job has been registered. You can execute the `getCronJob` function to view the details of your job played back. You can add multiple jobs on this contract in this fashion, and also use the other in your Job Scheduler to delete jobs or pause and unpause the Scheduler. However, you must register your Job Scheduler with the Chainlink Keepers network for it to be monitored and executed.
```

> You still have to register your Job Scheduler with the Chainlink Keepers service to have it monitored and executed.

Author: ```suggestion

```

## Setting up Chainlink Keepers

> Since our deployed contract is already [Keeper-compatible](../compatible-contracts), we can immediately register it with Keepers following these [steps](../register-upkeep). Be sure to provide the address of your Job Scheduler contract in **Upkeep address**. For Gas limit you should specify the upper limit of Gas your target function will use.

Author: ```suggestion
Since our deployed contract is already [Keeper-compatible](../compatible-contracts), we can immediately register it with Keepers following these [steps](../register-upkeep). Be sure to provide the address of your Job Scheduler contract in **Upkeep address**. For gas limits, you should specify the upper limit of gas your target function will use.
```

> 

Author: ```suggestion

> ❗️ Gas Limits
>
> The `KeeperRegistry` enforces a cap for gas used both on-chain and off-chain. See the [Keepers Network Overview](../overview/) for details. The caps are configurable and might change based on user feedback. Be sure that you understand these limits if your use case requires a large amount of gas.

```

Once you have registered your contract, you can view your Upkeep. The *Upkeep History* will display **Perform Upkeep** for all registered jobs on your Job Scheduler contract. To ensure Chainlink Keepers monitors your Job Scheduler, please ensure you fund your Upkeep.

# Summary

> In this tutorial you used the Chainlink Keepers Job Scheduler to automate the execution of your smart contract functions at specific times. 

Author: I think the Summary section is redundant and doesn't add to the walkthrough. Because it's only one sentence, I think we can get rid of it. Thoughts?

Reviewer: Yes the summary of where the user should end up when they complete the tutorial should be at the top as part of the overview.

# Encodings

> By using encodings you save the repeated gas costs of having your contract calculate them during execution.

Author: ```suggestion
Encodings allow you to save repeated gas costs of having your contract calculate them during execution.
```

Author: Is this sentence necessary to include here? It may be better to include when encodings are first introduced in the walkthrough

  +  [CronUpkeepFactory ABI](#cronupkeepfactory-abi)
  +  [Job Scheduler ABI](#job-scheduler-aka-cronupkeep-abi)
  +  [Contract Addresses](#contract-addresses)

# What is Job Scheduler?

The Chainlink Keepers Job Scheduler schedules jobs which utilize the Chainlink Keepers Network to automatically execute your smart contract functions at specific times. Once your Job Scheduler is registered with  the [Chainlink Keepers network](https://keepers.chain.link/new), the decentralized Keepers network will start monitoring and executing your scheduled jobs.  

# Job Scheduling Tutorial

We will use [Remix](https://remix.ethereum.org/) for this tutorial. To learn more about Remix and how you can use it to deploy a smart contract, visit our [tutorial](https://www.youtube.com/watch?v=JWJWT9cwFbo).

You will also need a MetaMask (or other Remix compatible wallet) in your browser with some testnet tokens of choice. See our [MetaMask tutorial](https://www.youtube.com/watch?v=4ZgFijd02Jo) if you need help setting up a wallet.

> ⚠️ Testing Contracts
> We recommend you first test the tutorial on an EVM-compatible testnet before moving to mainnet.

## Creating a Job Scheduler contract

> To deploy an on-chain Job Scheduler contract, we will use the externally audited and pre-deployed Chainlink  [CronUpkeepFactory](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/factories/CronUpkeepFactory.sol) smart contract. This is called the [Factory pattern](https://www.youtube.com/watch?v=Q1zZo4O_Ong), where one contract deploys another contract. To deploy your Job Scheduler (`CronUpkeep`), select the new file icon in Remix and name it `CronUpkeepFactory.abi`. Paste the [ABI](#cronupkeepfactory-abi) for `CronUpkeepFactory` from the [Resources](#resources) section below. The ABI is the application binary interface that is used to interact with the smart contract that has been deployed to the network.

Reviewer: Done

![Remix ABI Cron](/images/keepers/cron-1.png)

> Select the **DEPLOY & RUN TRANSACTIONS** icon in the left-hand navigation bar and set **Environment** to *Injected Web3*. Paste the address of our deployed `CronUpkeepFactory` contract in the **At Address** box and select the **At Address** button. You can find the [addresses](#contract-addresses) for deployed `CronUpkeepFactory` contracts in the [Resources](#resources) section. You will then be able to view callable functions of `CronUpkeepFactory`. Select the `newCronUpkeep` button and follow the prompts in MetaMask to confirm the transaction. When the transaction is complete,  a transaction hash will appear in the terminal. This points to the creation of your Job Scheduler.

Reviewer: No longer relevant with the new instructions

## Finding the address of your Job Scheduler

Copy the transaction hash and open it in the scanner of the chain you are working on. Once you have found it, select the **Internal Txns** tab. This displays the contract creation and contract address in the `To` field. Note the address and transaction hash for future use.

![Remix Transaction Hash](/images/keepers/cron-2.png)

## Interacting with your Job Scheduler

> To interact with your Job Scheduler contract, create a new file called `CronUpkeep.abi` in Remix and paste the [ABI](#job-scheduler-aka-cronupkeep-abi) for `CronUpkeep` contract from the [Resources](#resources) section below. Select the **DEPLOY & RUN TRANSACTIONS** icon on the left and paste the address of your Job Scheduler contract in the **At Address** box, and then select the **At Address** button. Accept the prompts after which the contract interface will open in the bottom left of Remix.

Reviewer: ```
To interact with your Job Scheduler contract,
```

Clarify that it's an ABI and not a contract. I went back in the docs looking to see which steps I missed because I thought there was a contract I didn't deploy.

Also clarify that when users deploy `CronUpkeep.abi`, they should deploy at the address for their `CronUpkeepFactory.abi` address in the Remix Deployed Contracts section.

Reviewer: Done.

![Remix Contract Interface](/images/keepers/cron-3.png)

## Scheduling a Job

To schedule a new job, use the function `createCronJobFromEncodedSpec`. This function has three inputs: 

> + The target address of the smart contract you want to automate

Reviewer: ```
+ The target address of the smart contract you want to automate
```
Can we give users an example contract to automate?

> + The [function signature](#function-signature), or the encoding of the function, you want to call 

Reviewer: ```
+ The [function signature](#function-signature), or the encoding of the function, you want to call 
```
If these signatures are not dynamic, can we provide a table of commonly used signatures? I'm not seeing any user input for the python script, so perhaps we provide it as an option but provide a list of signatures that we know will work.

Reviewer: No longer relevant. I misunderstood the function signatures.

> + The [encodedCronSpec](#encodedcronspec) encoding of your specified time interval.

Reviewer: ```
+ The [encodedCronSpec](#encodedcronspec) encoding of your specified time interval.
```

Give the user an example to copy and paste or find a way to automate the process of retrieving the encoded spec. That way users can just put in Cron format and get the result they want.

> Refer to the [Encodings](#encodings) section below to retrieve the [function signature](#function-signature) and the [encodedCronSpec](#encodedcronspec) for your job. 

Reviewer: ```
Refer to the [Encodings](#encodings) section below to retrieve the [function signature](#function-signature) and the [encodedCronSpec](#encodedcronspec) for your job. 
```

Reorganize this content to include each step in sequence so the user doesn't need to jump between sections of the document.

```json
<your target address>, <your function specification>, <your encodedCronSpec>
```
Replace the values and paste the snippet above into the `createCronJobFromEncodedSpec` function in Remix and execute the function. Once executed, you can call the `getActiveCronJobIDs` to see if your job has been registered. You can execute the `getCronJob` function to view the details of your job played back. You can add multiple jobs on this contract in this fashion, and also use the other in your Job Scheduler to delete jobs or pause and unpause the Scheduler. However, you must register your Job Scheduler with the Chainlink Keepers network for it to be monitored and executed.

## Setting up Chainlink Keepers

> Since our deployed contract is already [Keeper-compatible](../compatible-contracts), we can immediately register it with Keepers following these [steps](../register-upkeep). Be sure to provide the address of your Job Scheduler contract in **Upkeep address**. For gas limits, you should specify the upper limit of gas your target function will use.

Reviewer: ```suggestion
Because the deployed contract is already [Keeper-compatible](../compatible-contracts), you can immediately register it with Keepers following these [steps](../register-upkeep). Be sure to provide the address of your Job Scheduler contract in **Upkeep address**. For gas limits, you should specify the upper limit of gas your target function will use.
```

> Since our deployed contract is already [Keeper-compatible](../compatible-contracts), we can immediately register it with Keepers following these [steps](../register-upkeep). Be sure to provide the address of your Job Scheduler contract in **Upkeep address**. For gas limits, you should specify the upper limit of gas your target function will use.

Reviewer: ```
register it with Keepers following these [steps](../register-upkeep). 
```

Put the steps here. This is a full tutorial so we should provide the full process (other than requirements/setup steps) here if possible.

> ❗️ Gas Limits
>
> The `KeeperRegistry` enforces a cap for gas used both on-chain and off-chain. See the [Keepers Network Overview](../overview/) for details. The caps are configurable and might change based on user feedback. Be sure that you understand these limits if your use case requires a large amount of gas.

Once you have registered your contract, you can view your Upkeep. The *Upkeep History* will display **Perform Upkeep** for all registered jobs on your Job Scheduler contract. To ensure Chainlink Keepers monitors your Job Scheduler, please ensure you fund your Upkeep.

# Summary

In this tutorial you used the Chainlink Keepers Job Scheduler to automate the execution of your smart contract functions at specific times. 

# Encodings

> Encodings allow you to save repeated gas costs of having your contract calculate them during execution.

Reviewer: ```
Encodings allow you to save repeated gas costs of having your contract calculate them during execution.
```
Let's explain this a little more thoroughly. Why would the contract not just accept Cron in its normal format?
